feat: filter artifacts on --exclude-owned flag

thapabishwa opened the pull request:
## Description
This PR introduces the `--exclude-owned` flag to the Kubernetes resource handling. The purpose of this flag is to allow exclusion of Kubernetes objects that have an owner reference. This can be useful when we want to filter out resources that are managed or owned by other objects.

- Close #5064
- Depend on #5058 (#5019)

## Use cases
- Exclude Kubernetes artifacts that have owner references from being scanned. When `--exclude-owner` is  enabled, this flag streamlines scanning by skipping resources that are managed by higher-level controllers. 

### Before
```
trivy k8s pods,rs,statefulsets,deployments,daemonsets,jobs,cronjobs --report=summary --scanners config --all-namespaces
Workload Assessment
┌──────────────┬───────────────────────────────────────────────┬─────────────────────┐
│  Namespace   │                   Resource                    │  Misconfigurations  │
│              │                                               ├───┬───┬────┬────┬───┤
│              │                                               │ C │ H │ M  │ L  │ U │
├──────────────┼───────────────────────────────────────────────┼───┼───┼────┼────┼───┤
│ kube-system  │ Pod/ebs-csi-node-pjml9                        │   │ 1 │ 10 │ 20 │   │
│ kube-system  │ DaemonSet/ebs-csi-node                        │   │ 1 │ 10 │ 20 │   │
│ kube-system  │ Deployment/coredns                            │   │   │ 3  │ 5  │   │
│ kube-system  │ DaemonSet/aws-for-fluent-bit                  │   │   │ 4  │ 8  │   │
│ kube-system  │ ReplicaSet/coredns-66dddcb88c                 │   │   │ 3  │ 5  │   │
│ kube-system  │ DaemonSet/kube-proxy                          │   │ 2 │ 4  │ 10 │   │
│ kube-system  │ Deployment/ebs-csi-controller                 │   │   │ 5  │ 31 │   │
│ kube-system  │ Pod/aws-node-9f6mb                            │   │ 4 │ 8  │ 19 │   │
│ kube-system  │ DaemonSet/aws-node                            │   │ 2 │ 8  │ 20 │   │
│ kube-system  │ Pod/ebs-csi-controller-846b7ddddb-nj6ks       │   │   │ 5  │ 31 │   │
│ kube-system  │ Pod/ebs-csi-controller-846b7ddddb-tbl88       │   │   │ 5  │ 31 │   │
│ kube-system  │ Pod/kube-proxy-mp45r                          │   │ 2 │ 4  │ 10 │   │
│ kube-system  │ Pod/coredns-655c69d4f4-kmx6c                  │   │   │ 3  │ 5  │   │
│ kube-system  │ ReplicaSet/coredns-655c69d4f4                 │   │   │ 3  │ 5  │   │
│ kube-system  │ Pod/aws-for-fluent-bit-m46xg                  │   │   │ 4  │ 8  │   │
│ kube-system  │ ReplicaSet/ebs-csi-controller-846b7ddddb      │   │   │ 5  │ 31 │   │
│ kube-system  │ Pod/coredns-655c69d4f4-fg82d                  │   │   │ 3  │ 5  │   │
│ default      │ ReplicaSet/nginx-55875d95fd                   │   │   │ 3  │ 11 │   │
│ default      │ Pod/myteam-mytest-cluster5-1                  │   │   │ 3  │ 7  │   │
│ default      │ StatefulSet/myteam-mytest-cluster5            │   │   │ 3  │ 7  │   │
│ default      │ ReplicaSet/nginx-7dc58f44b5                   │   │   │ 3  │ 11 │   │
│ default      │ Pod/web-1                                     │   │   │ 3  │ 11 │   │
│ default      │ Pod/my-pod                                    │   │   │ 4  │ 11 │   │
│ default      │ Deployment/nginx                              │   │   │ 3  │ 11 │   │
│ default      │ Pod/frontend-82hnn                            │   │   │ 3  │ 11 │   │
│ default      │ ReplicaSet/frontend                           │   │   │ 3  │ 11 │   │
│ default      │ Deployment/postgres-operator-ui               │   │   │ 3  │ 7  │   │
│ default      │ Pod/nginx-5d56f7cc56-rj5tk                    │   │   │ 3  │ 11 │   │
│ default      │ Pod/myteam-mytest-cluster5-0                  │   │   │ 3  │ 7  │   │
│ default      │ ReplicaSet/nginx-5d56f7cc56                   │   │   │ 3  │ 11 │   │
│ default      │ ReplicaSet/postgres-operator-ui-7cf5759697    │   │   │ 3  │ 7  │   │
│ default      │ Pod/frontend-rkwww                            │   │   │ 3  │ 11 │   │
│ default      │ Pod/web-0                                     │   │   │ 3  │ 11 │   │
│ default      │ Pod/postgres-operator-5c884976b6-bbj7m        │   │   │ 1  │ 6  │   │
│ default      │ StatefulSet/web                               │   │   │ 3  │ 11 │   │
│ default      │ Pod/postgres-operator-ui-7cf5759697-hrlst     │   │   │ 3  │ 7  │   │
│ default      │ Pod/frontend-8jw4h                            │   │   │ 3  │ 11 │   │
│ default      │ ReplicaSet/postgres-operator-5c884976b6       │   │   │ 1  │ 6  │   │
│ default      │ Deployment/postgres-operator                  │   │   │ 1  │ 6  │   │
│ cert-manager │ ReplicaSet/cert-manager-cainjector-744bb89575 │   │   │ 1  │ 8  │   │
│ cert-manager │ Deployment/cert-manager                       │   │   │ 1  │ 8  │   │
│ cert-manager │ Pod/cert-manager-webhook-759d6dcbf7-6f4tc     │   │   │ 1  │ 8  │   │
│ cert-manager │ ReplicaSet/cert-manager-webhook-759d6dcbf7    │   │   │ 1  │ 8  │   │
│ cert-manager │ Deployment/cert-manager-webhook               │   │   │ 1  │ 8  │   │
│ cert-manager │ Deployment/cert-manager-cainjector            │   │   │ 1  │ 8  │   │
│ cert-manager │ Pod/cert-manager-cainjector-744bb89575-wrvs9  │   │   │ 1  │ 8  │   │
│ cert-manager │ Pod/cert-manager-69cdc85fc8-lmx5s             │   │   │ 1  │ 8  │   │
│ cert-manager │ ReplicaSet/cert-manager-69cdc85fc8            │   │   │ 1  │ 8  │   │
└──────────────┴───────────────────────────────────────────────┴───┴───┴────┴────┴───┘
Severities: C=CRITICAL H=HIGH M=MEDIUM L=LOW U=UNKNOWN
```
### After

```
trivy k8s pods,rs,statefulsets,deployments,daemonsets,jobs,cronjobs --report=summary --scanners config  --all-namespaces --exclude-owned

Workload Assessment
┌──────────────┬────────────────────────────────────┬─────────────────────┐
│  Namespace   │              Resource              │  Misconfigurations  │
│              │                                    ├───┬───┬────┬────┬───┤
│              │                                    │ C │ H │ M  │ L  │ U │
├──────────────┼────────────────────────────────────┼───┼───┼────┼────┼───┤
│ kube-system  │ DaemonSet/aws-node                 │   │ 2 │ 8  │ 20 │   │
│ kube-system  │ Deployment/coredns                 │   │   │ 3  │ 5  │   │
│ kube-system  │ DaemonSet/ebs-csi-node             │   │ 1 │ 10 │ 20 │   │
│ kube-system  │ Deployment/ebs-csi-controller      │   │   │ 5  │ 31 │   │
│ kube-system  │ DaemonSet/aws-for-fluent-bit       │   │   │ 4  │ 8  │   │
│ kube-system  │ DaemonSet/kube-proxy               │   │ 2 │ 4  │ 10 │   │
│ default      │ StatefulSet/web                    │   │   │ 3  │ 11 │   │
│ default      │ Deployment/nginx                   │   │   │ 3  │ 11 │   │
│ default      │ StatefulSet/myteam-mytest-cluster5 │   │   │ 3  │ 7  │   │
│ default      │ Deployment/postgres-operator-ui    │   │   │ 3  │ 7  │   │
│ default      │ Deployment/postgres-operator       │   │   │ 1  │ 6  │   │
│ default      │ Pod/my-pod                         │   │   │ 4  │ 11 │   │
│ default      │ ReplicaSet/frontend                │   │   │ 3  │ 11 │   │
│ cert-manager │ Deployment/cert-manager            │   │   │ 1  │ 8  │   │
│ cert-manager │ Deployment/cert-manager-webhook    │   │   │ 1  │ 8  │   │
│ cert-manager │ Deployment/cert-manager-cainjector │   │   │ 1  │ 8  │   │
└──────────────┴────────────────────────────────────┴───┴───┴────┴────┴───┘
Severities: C=CRITICAL H=HIGH M=MEDIUM L=LOW U=UNKNOWN
```
## Related PRs
- [x] #5043 

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aquasecurity/trivy?pullRequest=5059) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aquasecurity/trivy?pullRequest=5059) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aquasecurity/trivy?pullRequest=5059) it.</sub>

chen-keinan commented:
@thapabishwa the flag remain in trivy cli  however trivy should be aware to k8s internal notions for example : `ownerReferences`

one way to do it is to introduce `options` to client (in trivy-kubernetes project) see below:

```
type client struct {
	cluster       k8s.Cluster
	namespace     string
	resources     []string
	allNamespaces bool
	logger        *zap.SugaredLogger
	excludeOwned  bool
}

type K8sOption func(*client)

func WithExcludeOwned(excludeOwned bool) K8sOption {
	return func(c *client) {
		c.excludeOwned = excludeOwned
	}
}

// New creates a trivyK8S client
func New(cluster k8s.Cluster, logger *zap.SugaredLogger, opts ...K8sOption) TrivyK8S {
	c := &client{cluster: cluster, logger: logger}
	for _, opt := range opts {
		opt(c)
	}
	return c
}
```

thapabishwa commented:
> @thapabishwa the flag remain in trivy cli however trivy should be aware to k8s internal notions for example : `ownerReferences`
> 
> one way to do it is to introduce `options` pattern to client (in trivy-kubernetes project) see below:
> 
> ```
> type client struct {
> 	cluster       k8s.Cluster
> 	namespace     string
> 	resources     []string
> 	allNamespaces bool
> 	logger        *zap.SugaredLogger
> 	excludeOwned  bool
> }
> 
> type K8sOption func(*client)
> 
> func WithExcludeOwned(excludeOwned bool) K8sOption {
> 	return func(c *client) {
> 		c.excludeOwned = excludeOwned
> 	}
> }
> 
> // New creates a trivyK8S client
> func New(cluster k8s.Cluster, logger *zap.SugaredLogger, opts ...K8sOption) TrivyK8S {
> 	c := &client{cluster: cluster, logger: logger}
> 	for _, opt := range opts {
> 		opt(c)
> 	}
> 	return c
> }
> ```
> 
> once you have the flag in client its easy to include you functionality and also will help us in the future to introduce additional options


Thank you, @chen-keinan. 

I've understood the advantages of adding this changes into the `trivy-kubernetes` repository instead. I will send a PR to that repository and once it's approved I will then proceed to update this code once my work there is finished.

thapabishwa commented:
@chen-keinan 

This is the PR on `trivy-kubernetes` repo
https://github.com/aquasecurity/trivy-kubernetes/pull/214

chen-keinan commented:
> @chen-keinan
> 
> PR on `trivy-kubernetes` repo [aquasecurity/trivy-kubernetes#214](https://github.com/aquasecurity/trivy-kubernetes/pull/214)
> 
> Pulling it here as I cannot add reviewers on that PR.

@thapabishwa PR is approved and meged.
you can use this tag `v0.5.7-0.20230829110855-969eb077ad1c` to replace trivy-kubernetes in go mod file.

thapabishwa-plerionaut commented:
> > @chen-keinan
> > PR on `trivy-kubernetes` repo [aquasecurity/trivy-kubernetes#214](https://github.com/aquasecurity/trivy-kubernetes/pull/214)
> > Pulling it here as I cannot add reviewers on that PR.
> 
> @thapabishwa PR has been approved and merged. you can use this tag `v0.5.7-0.20230829110855-969eb077ad1c` to replace `trivy-kubernetes` in `go.mod` file.
> 
> _**Note:**_ you'll need to update cli docs after adding new flag :: use this command : `mage docs:generate`

Thanks @chen-keinan.

Upgraded  `trivy-kubernetes` to `v0.5.7-0.20230829110855-969eb077ad1c` and refactored existing code accordingly. 

Please let me know what you think about it.

chen-keinan commented on pkg/flag/kubernetes_flags.go:
```diff
@@ -79,6 +79,13 @@ var (
+		ConfigName: "kubernetes.exclude.owned",
+		Shorthand:  "E",
+		Default:    false,
+		Usage:      "exclude objects that have an owner reference",
```
maybe replace the wording `object` with `resource`

thapabishwa replied:
Done.

chen-keinan commented:
@thapabishwa FYI your PR is depend on #5058 as `trivy-kubernetes` dependency you used also picked up changes for #5050 pr , this the reason k8s integration test failing

thapabishwa commented on pkg/k8s/commands/resource.go:
```diff
@@ -22,11 +22,20 @@ func resourceRun(ctx context.Context, args []string, opts flag.Options, cluster
+		trivyk = trivyk8s.New(cluster, log.Logger, opts...)
+	} else {
+		trivyk = trivyk8s.New(cluster, log.Logger)
+	}
```
Done

chen-keinan commented:
> > @thapabishwa FYI your PR is depend on #5058 as `trivy-kubernetes` dependency you used also picked up changes for #5050 pr , this the reason k8s integration test failing
> 
> Thanks @chen-keinan . Appreciate your feedback. I've addressed to the comments you've made. Any suggestions to get the k8s integrations fixed so that this PR can be merged without depending on #5058 .

I'm do not think it can merged before #5058  as changes to trivy-kubernets lib wil impact trivy functionality , fixing the integration test will still keep trivy broken.
no worries merge will happen soon

thapabishwa commented:
> > > @thapabishwa FYI your PR is depend on #5058 as `trivy-kubernetes` dependency you used also picked up changes for #5050 pr , this the reason k8s integration test failing
> > 
> > 
> > Thanks @chen-keinan . Appreciate your feedback. I've addressed to the comments you've made. Any suggestions to get the k8s integrations fixed so that this PR can be merged without depending on #5058 .
> 
> I'm do not think it can be merged before #5058 as changes to trivy-kubernets lib will impact trivy k8s functionality , fixing the integration test will still keep trivy broken. no worries merge will happen soon :)


@chen-keinan I took some time to rethink the changes made to the `trivy-kubernetes` repository in https://github.com/aquasecurity/trivy-kubernetes/pull/214. 

In the current implementation, resources are skipped just if they have ownerRef. I think, this exclusion criteria might not be optimal as resources can be owned by custom resources.

As a consequence, I think it is more sensible to exclude only the resources that are owned by built-in workloads. I've created another PR in `trivy-kubernetes` to address the short comings https://github.com/aquasecurity/trivy-kubernetes/pull/215.

Please let me know what you think about it.

chen-keinan reviewed (dismissed):
@thapabishwa thank you for the contribution lgtm 🚀 .

lets wait for #5019 to get approved and merged (will also solve the integration test issue)

chen-keinan commented:
> > @thapabishwa thank you for the contribution lgtm 🚀 .
> > lets wait for #5019 to get approved and merged (will also solve the integration test issue)
> 
> Thanks @chen-keinan. Could you also take a look at [aquasecurity/trivy-kubernetes#215](https://github.com/aquasecurity/trivy-kubernetes/pull/215)? I think it might be beneficial for us to consider [aquasecurity/trivy-kubernetes#215](https://github.com/aquasecurity/trivy-kubernetes/pull/215) PR before merging the current one.

@thapabishwa  approved and merged , please update trivy-kubernetes version to `v0.5.7-0.20230830053006-95e88d51f82b` on go.mod file

thapabishwa commented:
> > > @thapabishwa thank you for the contribution lgtm 🚀 .
> > > lets wait for #5019 to get approved and merged (will also solve the integration test issue)
> > 
> > 
> > Thanks @chen-keinan. Could you also take a look at [aquasecurity/trivy-kubernetes#215](https://github.com/aquasecurity/trivy-kubernetes/pull/215)? I think it might be beneficial for us to consider [aquasecurity/trivy-kubernetes#215](https://github.com/aquasecurity/trivy-kubernetes/pull/215) PR before merging the current one.
> 
> @thapabishwa approved and merged , please update trivy-kubernetes version to `v0.5.7-0.20230830053006-95e88d51f82b` on go.mod file

@chen-keinan I realized that I made some mistake on the PR on `trivy-kubernetes`. Please take some time to look at this new PR that fixes this mistake.

https://github.com/aquasecurity/trivy-kubernetes/pull/216



> > > @thapabishwa thank you for the contribution lgtm 🚀 .
> > > lets wait for #5019 to get approved and merged (will also solve the integration test issue)
> > 
> > 
> > Thanks @chen-keinan. Could you also take a look at [aquasecurity/trivy-kubernetes#215](https://github.com/aquasecurity/trivy-kubernetes/pull/215)? I think it might be beneficial for us to consider [aquasecurity/trivy-kubernetes#215](https://github.com/aquasecurity/trivy-kubernetes/pull/215) PR before merging the current one.
> 
> @thapabishwa approved and merged , please update trivy-kubernetes version to `v0.5.7-0.20230830053006-95e88d51f82b` on go.mod file

Hey @chen-keinan I made some mistake on `trivy-kubernetes` library. 

Please take a look at this PR and let me know what you think about it. https://github.com/aquasecurity/trivy-kubernetes/pull/216

chen-keinan commented:
> > > > @thapabishwa thank you for the contribution lgtm 🚀 .
> > > > lets wait for #5019 to get approved and merged (will also solve the integration test issue)
> > > 
> > > 
> > > Thanks @chen-keinan. Could you also take a look at [aquasecurity/trivy-kubernetes#215](https://github.com/aquasecurity/trivy-kubernetes/pull/215)? I think it might be beneficial for us to consider [aquasecurity/trivy-kubernetes#215](https://github.com/aquasecurity/trivy-kubernetes/pull/215) PR before merging the current one.
> > 
> > 
> > @thapabishwa approved and merged , please update trivy-kubernetes version to `v0.5.7-0.20230830053006-95e88d51f82b` on go.mod file
> 
> Hey @chen-keinan. Sorry, I made some mistake on last PR to`trivy-kubernetes` library.
> 
> Please take a look at this PR and let me know what you think about it. [aquasecurity/trivy-kubernetes#216](https://github.com/aquasecurity/trivy-kubernetes/pull/216)

new tag for trivy-kubernetes: `v0.5.7-0.20230830063136-fe986af3f10f`

thapabishwa commented:
> v0.5.7-0.20230830063136-fe986af3f10f
@chen-keinan Thanks a lot. Appreciate your support and patience while I addressed the mistakes. It means a lot. Updated to new tag.

thapabishwa commented:
Rebased onto main to resolve `go.mod` and `go.sum` conflicts.

chen-keinan commented:
@thapabishwa Please rebase your branch with upstream and integration test issue will be fixed

thapabishwa commented:
> @thapabishwa Please rebase your branch with upstream and k8s integration test issue will be fixed

Thanks @chen-keinan . I've rebased the branch onto latest main.

knqyf263 commented on pkg/flag/kubernetes_flags.go:
```diff
@@ -79,6 +79,13 @@ var (
+	ExcludeOwned = Flag{
+		Name:       "exclude-owned",
+		ConfigName: "kubernetes.exclude.owned",
+		Shorthand:  "E",
```
We have a lot of flags now and want to assign shorthand to frequently used flags.

thapabishwa replied:
@chen-keinan Your thoughts on this?

thapabishwa commented:
@chen-keinan @knqyf263 . I've addressed to all the comments. Please let me know once this has been merged.

knqyf263 commented:
Thanks
